File: py/src/imbrace/resources/agent.py
from typing import Any, Dict
from ..http import HttpTransport, AsyncHttpTransport


def _with_defaults(body: Dict[str, Any]) -> Dict[str, Any]:
    """Auto-default empty provider_id/model_id on the assistant block.

    Without these, marketplace creates the assistant but ai-agent later
    returns 500 "Assistant is missing model_id/provider_id configuration"
    on chat. Mirrors the TS SDK behavior.
    """
    asst = dict(body.get("assistant") or {})
    if not asst.get("provider_id"):
        asst["provider_id"] = "system"
    if not asst.get("model_id"):
        asst["model_id"] = "gpt-4o"
    return {**body, "assistant": asst}


class AgentResource:
    """Agent / UseCase templates — Sync.

    Base URL is the marketplace service (`{gateway}/marketplaces/v2`).
    """

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        root = base.rstrip("/")
        self._templates = f"{root}/market-places/v2/templates"
        self._use_cases = f"{root}/use-cases"

    # ── Marketplace Templates ──────────────────────────────────────────────

    def list(self) -> Dict[str, Any]:
        return self._http.request("GET", self._templates).json()

    def list_agents(self) -> Dict[str, Any]:
        return self.list()

    def get(self, template_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._templates}/{template_id}").json()

    def get_agent(self, agent_id: str) -> Dict[str, Any]:
        return self.get(agent_id)

    def update(self, template_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PATCH", f"{self._templates}/{template_id}", json=body).json()

    def update_agent(self, agent_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self.update(agent_id, body)

    def delete(self, template_id: str) -> Dict[str, Any]:
        return self._http.request("DELETE", f"{self._templates}/{template_id}").json()

    def delete_agent(self, agent_id: str) -> Dict[str, Any]:
        return self.delete(agent_id)

    # ── Use-cases ──────────────────────────────────────────────────────────

    def list_use_cases(self) -> Dict[str, Any]:
        return self._http.request("GET", self._use_cases).json()

    def get_use_case(self, use_case_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._use_cases}/{use_case_id}").json()

    def create_use_case(self, body: Dict[str, Any]) -> Dict[str, Any]:
        """Create a use-case (atomic — usecase + assistant + workflow + channel).

        Expected body shape: `{"usecase": {"title": ...}, "assistant": {"name": ...}}`.
        SDK auto-fills `assistant.provider_id = "system"` and `assistant.model_id = "gpt-4o"`
        when empty/missing.
        """
        return self._http.request("POST", f"{self._use_cases}/v2/custom", json=_with_defaults(body)).json()

    def update_use_case(self, use_case_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PATCH", f"{self._use_cases}/{use_case_id}", json=body).json()

    def delete_use_case(self, use_case_id: str) -> Dict[str, Any]:
        return self._http.request("DELETE", f"{self._use_cases}/{use_case_id}").json()


class AsyncAgentResource:
    """Agent / UseCase templates — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        root = base.rstrip("/")
        self._templates = f"{root}/market-places/v2/templates"
        self._use_cases = f"{root}/use-cases"

    async def list(self) -> Dict[str, Any]:
        res = await self._http.request("GET", self._templates)
        return res.json()

    async def list_agents(self) -> Dict[str, Any]:
        return await self.list()

    async def get(self, template_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._templates}/{template_id}")
        return res.json()

    async def get_agent(self, agent_id: str) -> Dict[str, Any]:
        return await self.get(agent_id)

    async def update(self, template_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PATCH", f"{self._templates}/{template_id}", json=body)
        return res.json()

    async def update_agent(self, agent_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return await self.update(agent_id, body)

    async def delete(self, template_id: str) -> Dict[str, Any]:
        res = await self._http.request("DELETE", f"{self._templates}/{template_id}")
        return res.json()

    async def delete_agent(self, agent_id: str) -> Dict[str, Any]:
        return await self.delete(agent_id)

    async def list_use_cases(self) -> Dict[str, Any]:
        res = await self._http.request("GET", self._use_cases)
        return res.json()

    async def get_use_case(self, use_case_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._use_cases}/{use_case_id}")
        return res.json()

    async def create_use_case(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._use_cases}/v2/custom", json=_with_defaults(body))
        return res.json()

    async def update_use_case(self, use_case_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PATCH", f"{self._use_cases}/{use_case_id}", json=body)
        return res.json()

    async def delete_use_case(self, use_case_id: str) -> Dict[str, Any]:
        res = await self._http.request("DELETE", f"{self._use_cases}/{use_case_id}")
        return res.json()

File: ts/src/resources/agent.ts
import { HttpTransport } from "../http.js"
import type { AgentTemplate } from "../types/index.js"

export interface AgentAiAgentInput {
  name?: string
  description?: string
  model?: string
  instructions?: string
  [key: string]: unknown
}

export interface AgentUseCaseInput {
  name?: string
  description?: string
  category?: string
  [key: string]: unknown
}

export interface CreateAgentInput {
  /** Wire body key kept as `assistant` for backend compatibility. */
  assistant: AgentAiAgentInput
  usecase: AgentUseCaseInput
}

export interface UpdateAgentInput {
  /** Wire body key kept as `assistant` for backend compatibility. */
  assistant?: AgentAiAgentInput
  usecase?: AgentUseCaseInput
}

export interface DeleteAgentResponse {
  success: boolean
  [key: string]: unknown
}

export interface UseCase {
  _id: string
  name?: string
  description?: string
  category?: string
  [key: string]: unknown
}

/**
 * Wire body for `POST /use-cases/v2/custom` — must contain both `usecase`
 * (the marketplace template) and `assistant` (the AI agent paired with it).
 * Backend rejects payloads missing either key with 400 "missing usecase or
 * assistant payload".
 */
export interface CreateUseCaseInput {
  usecase: {
    title: string
    description?: string
    category?: string[] | string
    demo_url?: string
    [key: string]: unknown
  }
  assistant: {
    name: string
    /** Required by ai-service `/assistant_apps` — used to generate the underlying workflow. */
    workflow_name?: string
    description?: string
    instructions?: string
    model?: string
    model_id?: string
    provider_id?: string
    file_ids?: string[]
    mode?: string
    channel?: string
    category?: string[] | string
    metadata?: Record<string, unknown>
    [key: string]: unknown
  }
  [key: string]: unknown
}

export interface UpdateUseCaseInput {
  name?: string
  description?: string
  category?: string
  [key: string]: unknown
}

export class AgentResource {
  private readonly templates: string
  private readonly useCases: string

  /**
   * @param http  - HTTP transport
   * @param base  - marketplace service base URL (gateway/marketplaces/v2)
   */
  constructor(
    private readonly http: HttpTransport,
    base: string,
  ) {
    const root = base.replace(/\/$/, "")
    this.templates = `${root}/market-places/v2/templates`
    this.useCases  = `${root}/use-cases`
  }

  // ── Marketplace Templates ─────────────────────────────────────────────

  async list(): Promise<AgentTemplate[]> {
    return this.http.getFetch()(this.templates, { method: "GET" }).then(r => r.json())
  }

  async get(templateId: string): Promise<{ data: AgentTemplate }> {
    return this.http.getFetch()(`${this.templates}/${templateId}`, { method: "GET" }).then(r => r.json())
  }

  async update(templateId: string, body: UpdateAgentInput): Promise<AgentTemplate> {
    return this.http.getFetch()(`${this.templates}/${templateId}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async delete(templateId: string): Promise<DeleteAgentResponse> {
    return this.http.getFetch()(`${this.templates}/${templateId}`, { method: "DELETE" }).then(r => r.json())
  }

  // ── Use-cases ─────────────────────────────────────────────────────────

  async listUseCases(): Promise<UseCase[]> {
    return this.http.getFetch()(this.useCases, { method: "GET" }).then(r => r.json())
  }

  async getUseCase(useCaseId: string): Promise<UseCase> {
    return this.http.getFetch()(`${this.useCases}/${useCaseId}`, { method: "GET" }).then(r => r.json())
  }

  async createUseCase(body: CreateUseCaseInput): Promise<UseCase> {
    // Match `chatAi.createAiAgent` ergonomics — default empty/missing
    // model_id + provider_id to the org's system LLM (gpt-4o). Without
    // these, the assistant is created but ai-agent returns 500 on chat:
    // "Assistant is missing model_id/provider_id configuration".
    const wireBody: CreateUseCaseInput = {
      ...body,
      assistant: {
        ...body.assistant,
        provider_id: body.assistant?.provider_id || "system",
        model_id:    body.assistant?.model_id    || "gpt-4o",
      },
    }
    return this.http.getFetch()(`${this.useCases}/v2/custom`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(wireBody),
    }).then(r => r.json())
  }

  async updateUseCase(useCaseId: string, body: UpdateUseCaseInput): Promise<UseCase> {
    return this.http.getFetch()(`${this.useCases}/${useCaseId}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteUseCase(useCaseId: string): Promise<DeleteAgentResponse> {
    return this.http.getFetch()(`${this.useCases}/${useCaseId}`, { method: "DELETE" }).then(r => r.json())
  }
}
